src/monty.rs

```diff
@@ -1,129 +1,221 @@
-    // TODO: Replace with an optimised squaring function
-    monty_redc(&a * &a, mr)
+/// z1 << _W + z0 = x * y + c
+#[inline(always)]
+fn mul_add_www(x: BigDigit, y: BigDigit, c: BigDigit) -> (BigDigit, BigDigit) {
+    let z = x as DoubleBigDigit * y as DoubleBigDigit + c as DoubleBigDigit;
+    ((z >> big_digit::BITS) as BigDigit, z as BigDigit)
 }
 
-pub fn monty_modpow(a: &BigUint, exp: &BigUint, modulus: &BigUint) -> BigUint {
-    let mr = MontyReducer::new(modulus);
+/// Calculates x ** y mod m using a fixed, 4-bit window.
+pub fn monty_modpow(x: &BigUint, y: &BigUint, m: &BigUint) -> BigUint {
+    assert!(m.data[0] & 1 == 1);
+    let mr = MontyReducer::new(m);
+    let num_words = m.data.len();
 
-    // Calculate the Montgomery parameter
-    let mut v = vec![0; modulus.data.len()];
-    v.push(1);
-    let r = BigUint::new(v);
+    let mut x = x.clone();
+
+    // We want the lengths of x and m to be equal.
+    // It is OK if x >= m as long as len(x) == len(m).
+    if x.data.len() > num_words {
+        x %= m;
+        // Note: now len(x) <= numWords, not guaranteed ==.
+    }
+    if x.data.len() < num_words {
+        x.data.resize(num_words, 0);
+    }
 
-    // Map the base to the Montgomery domain
-    let mut apri = a * &r % modulus;
+    // rr = 2**(2*_W*len(m)) mod m
+    let mut rr = BigUint::one();
+    rr = (rr.shl(2 * num_words * big_digit::BITS)) % m;
+    if rr.data.len() < num_words {
+        rr.data.resize(num_words, 0);
+    }
+    // one = 1, with equal length to that of m
+    let mut one = BigUint::one();
+    one.data.resize(num_words, 0);
+
+    let n = 4;
+    // powers[i] contains x^i
+    let mut powers = Vec::with_capacity(1 << n);
+    powers.push(montgomery(&one, &rr, m, mr.n0inv, num_words));
+    powers.push(montgomery(&x, &rr, m, mr.n0inv, num_words));
+    for i in 2..1 << n {
+        let r = montgomery(&powers[i - 1], &powers[1], m, mr.n0inv, num_words);
+        powers.push(r);
+    }
+
+    // initialize z = 1 (Montgomery 1)
+    let mut z = powers[0].clone();
+    z.data.resize(num_words, 0);
+    let mut zz = BigUint::zero();
+    zz.data.resize(num_words, 0);
+
+    // same windowed exponent, but with Montgomery multiplications
+    for i in (0..y.data.len()).rev() {
+        let mut yi = y.data[i];
+        let mut j = 0;
+        while j < big_digit::BITS {
+            if i != y.data.len() - 1 || j != 0 {
+                zz = montgomery(&z, &z, m, mr.n0inv, num_words);
+                z = montgomery(&zz, &zz, m, mr.n0inv, num_words);
+                zz = montgomery(&z, &z, m, mr.n0inv, num_words);
+                z = montgomery(&zz, &zz, m, mr.n0inv, num_words);
+            }
+            zz = montgomery(
+                &z,
+                &powers[(yi >> (big_digit::BITS - n)) as usize],
+                m,
+                mr.n0inv,
+                num_words,
+            );
+            ::std::mem::swap(&mut z, &mut zz);
```

Comment:
what's up with the preceding namespace qualifier `::`? I didn't know you could do this, and it seems unnecessary. Wouldn't `std::mem::swap` work?

Reply:
You can only do that since [Rust 1.30](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1300-2018-10-25):

> Using a external crate no longer requires being prefixed with `::`.

But anyway I prefer to `use std::mem` at the top, then call `mem::swap`. Will change.